lib/confluence/aggressive_removal.es6.js

```diff
@@ -43,97 +39,110 @@ foam.CLASS({
       documentation: `Compute aggressive removal value for each browser in
           browsers at the given date.`,
       code: function(browsers, date) {
-        let computeForBrowsers = [];
-        for (let i = 0; i < browsers.length; i++) {
-          let browser = browsers[i];
-          let year = date.getFullYear();
-          let month = date.getMonth();
-          let day = date.getDate();
-          let oneYearAgo = new Date(year - 1, month, day);
-          let browserKeys = browsers.map((browser) => browser.browserKey);
-          let selectBrowsers = [
-            this.browserDAO.where(this.mlang.AND(
-                this.mlang.NEQ(this.Browser.BROWSER_NAME, browser.browserName),
-                this.mlang.IN(this.Browser.BROWSER_KEY, browserKeys))).select(),
-            this.browserDAO.where(this.mlang.AND(
-                this.mlang.EQ(this.Browser.BROWSER_NAME, browser.browserName),
-                this.mlang.LT(this.Browser.RELEASE_DATE, oneYearAgo),
-                this.mlang.EQ(this.Browser.OS_NAME, browser.osName))).select()];
-          computeForBrowsers.push(Promise.all(selectBrowsers).then((result) => {
-            let currBrowsers = result[0].a;
-            let prevReleaseBrowsers = result[1].a;
-            // Return if there is insufficient of historical data. This method
-            // returns Promise<Sink> (result of select() call).
-            if (prevReleaseBrowsers.length < 2)
-              return Promise.resolve(this.ArraySink.create());
+        return Promise.all(browsers.map(
+            browser => this.computeForBrowser(browsers, date, browser)));
+      },
+    },
+    {
+      name: 'computeForBrowser',
+      documentation: 'Compute aggressive removal value for a single browser.',
+      code: function(browsers, date, browser) {
+        let browserKeys = browsers.map((browser) => browser.browserKey);
+        return Promise.all([
+          // Current browsers from vendor other than "browser".
+          this.browserDAO.where(this.mlang.AND(
+              this.mlang.NEQ(this.Browser.BROWSER_NAME, browser.browserName),
+              this.mlang.IN(this.Browser.BROWSER_KEY, browserKeys)))
+                  .orderBy(this.mlang.DESC(this.Browser.RELEASE_DATE))
```

Comment:
How does this produce just the latest version for each browser key? Is it missing a group by somewhere?

browserDAO has a function getLatestBrowserFromEachVendorAtDate. Is it possible to use that here?